src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java

```diff
@@ -125,7 +126,7 @@ public void onEvent(ResourceOffersEvent event, long sequence,
     private boolean matches(Offer offer, NMProfile profile) {
         double cpus = -1;
```

Comment:
As a general comment, and even though we do not use the PMD codesize option (maybe we should) this method has exceeded the accepted standards for Cyclomatic Complexity. The method matches() has a complexity of 18 and a  NpathComplexity of over 200. The general accepted CC level is 5-7, though some feel 10 should be the limit. The output of the codesize PMD option looks like:

9 myriad/src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java 	50 	The class 'ResourceOffersEventHandler' has a Cyclomatic Complexity of 14 (Highest = 18).
10 	myriad/src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java 	126 	The method 'matches' has a Cyclomatic Complexity of 18.
11 	myriad/src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java 	126 	The method matches() has an NPath complexity of 440

I took the liberty of refactoring the code as an example on how to reduce the complexity, and remove the PMD issues. Though this increased the overall complexity of the class it reduced the complexity per method which would simplify, in theory, unit test, code coverage, and the general quality of the code. The following is the refactor:

```
 private boolean matches(Offer offer, NMProfile profile) {
        Map<String, Object> results = new HashMap<String, Object>(5);
        
        for (Resource resource : offer.getResourcesList()) {
            if (resourceEvaluators.containsKey(resource.getName())) {
                resourceEvaluators.get(resource.getName()).eval(resource, results);
            } else {
                LOGGER.warn("Ignoring unknown resource type: {}",
                        resource.getName());
            }
        }
        double cpus = (Double) results.get("cpus");
        double mem = (Double) results.get("mem");
        int ports = (Integer) results.get("ports");
        
        checkResource(cpus < 0, "cpus");
        checkResource(mem < 0, "mem");
        checkResource(ports < 0, "port");
        
        return checkAggregates(offer, profile, ports, cpus, mem);
    }
    
    private void checkResource(boolean fail, String resource) {
        if (fail) {
            LOGGER.error("No " + resource + " resources present");
        }
    }
    
    private boolean checkAggregates(Offer offer, NMProfile profile, int ports, double cpus, double mem) {
        Map<String, String> requestAttributes = new HashMap<>();

        if (taskUtils.getAggregateCpus(profile) <= cpus
                && taskUtils.getAggregateMemory(profile) <= mem
                && SchedulerUtils.isMatchSlaveAttributes(offer, requestAttributes)
                && NMPorts.expectedNumPorts() <= ports) {
            return true;
        } else {
            LOGGER.info("Offer not sufficient for task with, cpu: {}, memory: {}, ports: {}",
                    taskUtils.getAggregateCpus(profile), taskUtils.getAggregateMemory(profile), ports);
            return false;
        }      
    }
    
    private static Double scalarToDouble(Resource resource, String id) {
        Double value = new Double(0.0);
        if (resource.getType().equals(Value.Type.SCALAR)) {
            value = new Double(resource.getScalar().getValue());
        } else {
            LOGGER.error(id + " resource was not a scalar: {}", resource
                    .getType().toString());
        }
        return value;
    }
    
    private interface EvalResources {
        public void eval(Resource resource, Map<String, Object>results);
    }
    
    private static Map<String, EvalResources> resourceEvaluators;
    
    static {
        resourceEvaluators = new HashMap<String, EvalResources>(4);
        resourceEvaluators.put("cpus", new EvalResources() {
            public void eval(Resource resource, Map<String, Object> results) {
                results.put("cpus", scalarToDouble(resource, "cpus"));
            }
        });
        resourceEvaluators.put("mem", new EvalResources() {
            public void eval(Resource resource, Map<String, Object> results) {
                results.put("mem", scalarToDouble(resource, "mem"));
            }
        });
        resourceEvaluators.put("disk", new EvalResources() {
            public void eval(Resource resource, Map<String, Object> results) {
            }
        }); 
        resourceEvaluators.put("ports", new EvalResources() {
            public void eval(Resource resource, Map<String, Object> results) {
                int ports = 0;
                if (resource.getType().equals(Value.Type.RANGES)) {
                    Value.Ranges ranges =  resource.getRanges();
                    for (Value.Range range : ranges.getRangeList()) {
                        if (range.getBegin() < range.getEnd()) {
                            ports += range.getEnd() - range.getBegin() + 1;
                        }
                    }
                    
                } else {
                    LOGGER.error("ports resource was not Ranges: {}", resource
                            .getType().toString());
                    
                }
                results.put("ports", Integer.valueOf(ports));
            }
        });
        
    }
    
```